### ISSUE TITLE
igluctl: consider master static command

### ISSUE THREAD

**Issue by chuwy** [snowplow/iglu#375]
Command that performs whole usual workflow as single step:

1. `lint`
2. `static generate --with-json-paths`
3. `static push` to Iglu Server (possibly several)
4. `static s3cp` schemas to Static Registry (possibly several)
5. `static s3cp` JSONPaths to S3
6. Optional: deploy new table (new subcommand)

Command can short-circuit on first failed step (e.g. lint is not good enough) and be idempotent.

**Comment by colmsnowplow** [comment]
Awesome idea!

A couple of notes/things to consider:

- We might consider another `master-test` command which pushes to SP Mini - in my experience people will generally follow the path of least resistance with this tool, so there's a risk of people pushing untested schemas if we make that easier than testing.

- The `lint` portion of the command should default to skip the description check - a lot of users forgo the field descriptions, and really that isn't a necessity for a schema. It could be quite a frustrating UX if they're blocked by this.

- On the above point but not strictly this use case: I think the `lint` command should default to skip the description check when run individually too - for the same reasons (let me know if that should be a discussion elsewhere/another issue).

**Comment by alexanderdean** [comment]
I think the idea here is great but as @colmsnowplow says, we need to be careful about blurring the dev-test cycle and the prod cycle into one.

**Comment by chuwy** [comment]
Hey @colmsnowplow, all your points are actually covered by idea of configuration file:

* Users might want to have separate config for Snowplow Mini, where only Mini's address specified in `registries` array
* Config file would have array of enabled/disabled checks
* ...which means individual file will be able to disable `description`-check

Basically user will have two separate config files `production.json` and `mini.json`

**Comment by chuwy** [comment]
Scheduled this for R10 `static deploy`. Please feel free to advice on better name if someone have any suggestions.

**Comment by keanerobinson** [comment]
Fully agree with not blurring the testing / dev cycle into one! Some other ideas:

- One thing that'd be nice on the QA side would be to automate the clearing of the cache when updating existing schemas. It's quite easy to forget to do this when you're rolling out tracking and it can be quite confusing when you expect the events to land in Mini.

- When using `static generate --with-json-paths` it would make sense if the command also deleted the pre-existing jsonpath file and replaced it with the new one. If you don't delete the file manually you get a message which says that the file already exists.

**Comment by colmsnowplow** [comment]
@keanerobinson 

> When using static generate --with-json-paths it would make sense if the command also deleted the pre-existing jsonpath file and replaced it with the new one. If you don't delete the file manually you get a message which says that the file already exists.

This functionality exists with the --force argument I believe (which overwrites the files). Agree that it should be taken into consideration in the next version.

**Comment by keanerobinson** [comment]
good point - I've gotten so used to just manually deleting the file beforehand. It feels like `--force` should be run by default for the jsonpaths or be set in the mentioned config filea.

**Comment by chuwy** [comment]
Hey @keanerobinson! Thanks for the feedback! But unfortunately I cannot do anything regarding both issues you've raised.

1. Igluctl knows nothing about Snowplow mini and its cache. It works with Iglu Server directly which is an independent Snowplow Mini component. To be honest, I understand how annoying it can be, but cannot even think of what component can be responsible for automatic cache clearing. Probably something in Snowplow Insights, but we're far away from it
2. `--force` by default is a very dangerous behavior. We'll break much more things if Igluctl will be able to mutate existing assets for any smallest change. But on a very bright side - we should start working on removing JSONPaths altogether so a lot of related issues should go away with it.

**Comment by chuwy** [comment]
> or be set in the mentioned config files

Although, this can be an option!

**Comment by keanerobinson** [comment]
**re 1:** good to know! 
**re 2:** Looking forward to not having to worry about jsonpaths! Being able to set `with-json-paths` to delete and replace existing jsonpaths in the config file would also still be a time saver.

One final note on the proposed `master-test`. I think this command should:

- `lint`
- `static push`

I'm not sure if it makes sense to static generate the DDL / jsonpaths at this stage as a QA session with SP Mini often entails further schema changes. Again, this could also be a configurable command.

**Comment by alexanderdean** [comment]
> what component can be responsible for automatic cache clearing

Probably a Snowplow Control Plane API call eventually.

**Comment by chuwy** [comment]
> Probably a Snowplow Control Plane API call eventually.

Sure, but I mean on client-side

**Comment by chuwy** [comment]
Few points on configuration file (leaving here as https://github.com/snowplow/iglu/issues/383 will likely be merged into this ticket):

* As @colmsnowplow mentioned - we need to consider igluctl configuration file as a definition of particular deploy process. However user would like to separate his Test/Mini and Prod environments should be reflected in two separate configs
* Therefore, as @keanerobinson proposed - it also should include list of commands. E.g. a separate array that in `igluctl_mini.json` would contain  `"steps": ["lint", "static push"]` and without any S3 buckets for JSONPaths and without options specific to `static generate`
* In `igluctl lint` we decided to *skip* checks because this is how users could always have strictest possible linting by default. But in `static deploy` it could be other way round (*include*) because configuration files could be just copied around and users wouldn't have to remember all possible checks and add them every time

**Comment by alexanderdean** [comment]
> in static deploy it could be other way round (include)

This makes sense because then a deploy process doesn't suddenly break because of an Igluctl upgrade (because the existence of a new check won't matter until config files are updated to make use of it).

**Comment by alexanderdean** [comment]
Can we set out exactly how the new functionality will be used? An example would be ideal. Reading this thread to try to understand that is quite confusing...
